lib/staccato/v4/tracker.rb

```diff
@@ -1,155 +1,334 @@
+    def add_adapter(adapter, logger = nil, formatter = nil)
+      case adapter
+      when :logger
+        require 'staccato/adapter/logger'
+        uri = self.class.collection_uri
+        adapter = Staccato::Adapter::Logger.new(uri, logger, formatter)
+      when :net_http
+        adapter = self.class.default_adapter
+      when :validation
+        adapter = self.class.validation_adapter
+      end
       @adapters << adapter
+      adapter
     end
 
-    # The unique client id
+    # The identifier for a Data Stream provided by google, i.e., `G-XXXXXXXX`
     # @return [String]
-    def client_id
-      @client_id ||= Staccato.build_client_id
+    attr_reader :measurement_id
+
+    # The required API Secret that is generated through the Google Analytics UI
+    # @return [String]
+    attr_reader :api_secret
+
+    # uniquely identifies a user instance of a web client
+    # @return [String]
+    attr_reader :client_id
+
+    # An optional unique identifier for a user.  Optional.
+    # @return [String, nil]
+    attr_reader :user_id
+
+    # User properties describe segments of your user base, such as language
+    # preference or geographic location.  Optional.
+    # @return [Hash]
+    attr_reader :user_properties
+
+    # Indicates that these events should not be used for personalized ads
+    # @return [Boolean]
+    attr_reader :non_personalized_ads
+
+    # The time to associate with events.  This should only be set to record
+    # events that happened in the past.  This value can be overridden via
+    # user_property or event timestamps.
+    #
+    # When set, this returns the Time associated with {#timestamp_micros}.
+    #
+    # See {Staccato::V4.timestamp_micros}.
+    #
+    # @return [Time, nil]
+    attr_reader :time
+
+    # The time to associate with the events, as a Unix timestamp in
+    # microseconds.  It is recommended to use {#time} instead.
+    # @return (see #time)
+    attr_reader :timestamp_micros
+
+    # The default event parameters used in all events sent from this tracker
+    # @return [Hash]
+    attr_reader :default_event_params
+
+    # Use {#<<}, {#add}, {#clear}, or {#trim} to update the events.  The array
+    # is a frozen duplicate, which cannot be updated directly.
+    #
+    # Note: Requests can have a maximum of 25 events each.  When more than 25
+    # events are present, multiple HTTP requests will be sent.
+    #
+    # @return [Array[Staccato::V4::Event]] the unsent events
+    def events; @events.dup.freeze end
+
+    # Add an event to the events to be sent, and return self.
+    #
+    # @param time [Time] The time to associate with the event.  This should only
+    #     be set to record events that happened in the past.  It will be
+    #     translated into a timestamp_micros event parameter.
+    # @param params [Time] event parameters to me used with the event
+    # @return self
+    def add(event, time: nil, **params)
```

Comment:
These changes are my favorite part of the PR. I think this is a great addition to support "custom" events.